Allow to set a max bound for pooled memory in PooledByteBufAllocator.…

normanmaurer opened the pull request:
… Releated to [#4471]

Motivation:

Sometimes it is useful to set a max cap on the amount of memory that will be pooled in PooledByteBufAllocator and after this is reached just allocate
unpooled memory.

Modifications:

Allow the maximum of pooled memory after which unpooled memory is used.

Result:

Its possible to cap the maximum of pooled memory for the allocator.

normanmaurer commented:
@trustin PTAL

trustin commented on buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:
```diff
@@ -103,6 +104,8 @@
         DEFAULT_CACHE_TRIM_INTERVAL = SystemPropertyUtil.getInt(
                 "io.netty.allocator.cacheTrimInterval", 8192);
 
+        DEFAULT_MAX_POOLED_BYTES = SystemPropertyUtil.getLong("io.netty.allocator.maxPooledBytes", Integer.MAX_VALUE);
```
`Long.MAX_VALUE`?

normanmaurer replied:
oh jeeez yes!

trustin commented on buffer/src/main/java/io/netty/buffer/PoolArena.java:
```diff
@@ -120,6 +123,7 @@ protected PoolArena(PooledByteBufAllocator parent, int pageSize, int maxOrder, i
         metrics.add(q075);
         metrics.add(q100);
         chunkListMetrics = Collections.unmodifiableList(metrics);
+        this.maxBytesPooled = Math.min(maxBytesPooled, chunkSize);
```
Shouldn't this be `max`?

normanmaurer replied:
yes... I need more coffee..

trustin commented:
Could you add a test case for this?

normanmaurer commented:
@trustin yep working on it :)

netkins commented on buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:
```diff
@@ -156,13 +160,22 @@ public PooledByteBufAllocator(boolean preferDirect, int nHeapArena, int nDirectA
+        this(preferDirect, nHeapArena, nDirectArena, pageSize, maxOrder,
+                tinyCacheSize, smallCacheSize, normalCacheSize, DEFAULT_MAX_POOLED_BYTES);
+    }
+    public PooledByteBufAllocator(boolean preferDirect, int nHeapArena, int nDirectArena, int pageSize, int maxOrder,
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Constructor has 9 parameters, which is greater than 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS00107)

normanmaurer commented:
@trustin I added a test and addressed your comments but wondered about something.

At the moment the limit is per arena, not sure this is good enough or not. Do we need to allow to use different limits per type (like one for heap and one for direct buffers) ? Or do we need to allow a "global" limit ?

blucas commented on buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:
```diff
@@ -103,6 +104,8 @@
         DEFAULT_CACHE_TRIM_INTERVAL = SystemPropertyUtil.getInt(
                 "io.netty.allocator.cacheTrimInterval", 8192);
 
+        DEFAULT_MAX_POOLED_BYTES = SystemPropertyUtil.getLong("io.netty.allocator.maxPooledBytes", Long.MAX_VALUE);
```
If this setting is per arena, then I think the system property should be named accordingly (such as ```io.netty.allocator.arena.maxPooledBytes```), otherwise this property gives the impression that the limit is for the entire allocator.

normanmaurer replied:
yep... that said I'm debating if it should be per arena or not.. WDYT ?

johnou replied:
I'm for io.netty.allocator.maxPooledBytes if it represents max pooled bytes for all arenas.

trustin commented:
+1 for global limit rather than per-arena limit.

normanmaurer commented:
@blucas @trustin PTAL again.

buchgr commented:
If the limit is reached, the allocation still succeeds but the memory is not pooled, right? I think in that case what you really want is a better deallocation strategy that adapts to different allocation / deallocation behaviour. The code change to enforce such a limit is easy, but it's really hard and complex for a user to set a good limit on memory, as this depends largely on the load and code paths taken. In many applications this might change over time. I think introducing such a limit, might often lead to unpredictable performance degradations.

normanmaurer commented:
@buchgr your understanding is correct. I agree it might be hard for someone to set a good limit, that said I think it is still useful.

johnou commented:
@buchgr handy for test / staging environments with tight resources.

buchgr commented:
@johnou I disagree. A decent dirty page purging mechanism will solve this, without having to set some max memory.

buchgr commented:
@johnou One problem with a global memory limit is, that memory is still allocated / deallocated per arena. Say you have two arenas. Your global limit is 1GB. Arena 1 has 300MB allocated and none-cached, Arena 2 has 700MB allocated, thereof 200MB are unused. Now when you make an allocation into Arena 1, it will fallback to requesting unpooled memory. However, your purging mechanism works, you will never hit that limit. The global limit doesn't magically limit the memory your application uses.

johnou commented:
@buchgr I wasn't saying that your idea of a purging mechanism is a bad idea just that being able to set the global memory limit instead of trying to figure out a suitable arena count, tinyCacheSize / smallCacheSize and normalCacheSize or disabling pooled buffers altogether is easier to manage.

blucas commented:
@normanmaurer - I'm with @buchgr on this one. I think a per-arena limit is sufficient. Users can control the number of arenas created, so they can determine the global (pooled) memory limit.

johnou commented:
@blucas @buchgr when I first read global memory limit I thought limiting would be done per arena but the size per area would be calculated by global memory pool / arena count.

blucas commented on buffer/src/main/java/io/netty/buffer/PoolArena.java:
```diff
@@ -225,21 +225,28 @@ private void allocate(PoolThreadCache cache, PooledByteBuf<T> buf, final int req
-            return;
-        }
+    private void allocateNormal(PooledByteBuf<T> buf, int reqCapacity, int normCapacity) {
+        synchronized (this) {
```
Just for my understanding, you removed ```synchronized``` from the method declaration and introduced this ```synchronized block```. Is this because you realized that the call to ```allocateHuge()``` later in this method is not required to be within a ```synchronized``` portion of code?

buchgr commented:
@johnou 
> I wasn't saying that your idea of a purging mechanism is a bad idea 

That's not my idea :-). That's how jemalloc does it.

> just that being able to set the global memory limit instead of trying to figure out a suitable arena count, tinyCacheSize / smallCacheSize and normalCacheSize or disabling pooled buffers altogether is easier to manage.

I don't see how the global memory limit relates to those settings? It doesn't spare you from finding appropriate values.

trustin commented:
Hmmmm, on a second thought, I'm not sure what the benefit of having this feature is. Could you explain why this is useful?

normanmaurer commented:
Actually I think @buchgr is correct and we should better work on the "pruning" strategy. Closing...
